File: nattybear/level2/문자열-압축/solution.js
function solution(s) {
  if (s.length == 1)
    return 1
  const xs = []
  for (let i = 1; i < s.length; i++)
    xs.push(compress(s, i).length)
  return Math.min(...xs)
}

function compress(s, n) {
  return split(s, n)
    .map(count)
    .join('')
}

function split(s, n) {
  let can = []
  const bag = []
  function go(s) {
    if (s.length == 0)
      return
    const fst = take(n, s)
    const tail = drop(n, s)
    const snd = take(n, tail)
    if (fst == snd) {
      can.push(fst)
      go(tail)
    }
    else {
      can.push(fst)
      bag.push(can)
      can = []
      go(tail)
    }
  }
  go(s)
  return bag
}

function count(xs) {
  const x = take(1, xs)
  const n = xs.length
  if (n == 1)
    return x
  else
    return n.toString() + x
}

function take(n, s) {
  if (s.length <= n)
    return s
  return s.slice(0, n)
}

function drop(n, s) {
  if (s.length <= n)
    return ""
  return s.slice(n)
}

File: nattybear/level2/문자열-압축/Solution.java
import java.util.*;
import java.util.stream.*;

class Solution {
  public int solution(String s) {
    if (s.length() == 1)
      return 1;
    ArrayList<Integer> list = new ArrayList<>();
    Robot robot;
    for (int i = 1; i < s.length(); i++) {
      robot = new Robot(i);
      list.add(robot.compress(s).length());
    }
    return Collections.min(list);
  }
}

class Car extends ArrayList<char[]> {}

class Train extends ArrayList<Car> {}

class Robot {
  private int n;
  private Car car;
  private Train train;

  public Robot(int n) {
    this.n = n;
    car = new Car();
    train = new Train();
  }

  public String compress(String s) {
    split(s.toCharArray());
    return train 
      .stream()
      .map(x -> count(x))
      .collect(Collectors.joining());
  }

  public void split(char[] s) {
    if (s.length == 0)
      return;
    char[] fst = Util.take(n, s);
    char[] tail = Util.drop(n, s);
    char[] snd = Util.take(n, tail);
    if (Arrays.equals(fst, snd)) {
      car.add(fst);
      split(tail);
    } else {
      car.add(fst);
      train.add(car);
      car = new Car();
      split(tail);
    }
  }

  public String count(Car car) {
    char[] c = car.get(0);
    int len = car.size();
    if (len == 1)
      return String.valueOf(c);
    return String.valueOf(len) + String.valueOf(c);
  }

  public Train getTrain() {
    return train;
  }
}

class Util {
  public static char[] take(int n, char[] s) {
    if (s.length <= n)
      return s;
    char[] t = new char[n];
    for (int i = 0; i < n; i++)
      t[i] = s[i];
    return t;
  }

  public static char[] drop(int n, char[] s) {
    if (s.length <= n)
      return new char[] {};
    int newLength = s.length - n;
    char[] t = new char[newLength];
    int j = n;
    for (int i = 0; i < newLength; i++)
      t[i] = s[j++];
    return t;
  }
}
